programs/08-wasm-eth/src/contract.rs

```diff
@@ -0,0 +1,649 @@
+    let wasm_consensus_state = WasmConsensusState {
+        data: consensus_state.clone().into(),
+    };
+    let wasm_consensus_state_any = Any::from_msg(&wasm_consensus_state).unwrap();
+    let height = Height {
+        revision_number: 0,
+        revision_height: consensus_state.slot,
+    };
+    deps.storage.set(
+        consensus_db_key(&height).as_bytes(),
+        wasm_consensus_state_any.encode_to_vec().as_slice(),
+    );
+
+    Ok(Response::default())
+}
+
+#[entry_point]
+pub fn sudo(
+    deps: DepsMut<EthereumCustomQuery>,
+    _env: Env,
+    msg: SudoMsg,
+) -> Result<Response, ContractError> {
+    let result = match msg {
+        SudoMsg::VerifyMembership(verify_membership_msg) => {
+            verify_membership(deps.as_ref(), verify_membership_msg)?
+        }
+        SudoMsg::VerifyNonMembership(verify_non_membership_msg) => {
+            verify_non_membership(deps.as_ref(), verify_non_membership_msg)?
+        }
+        SudoMsg::UpdateState(_) => update_state()?,
+        SudoMsg::UpdateStateOnMisbehaviour(_) => unimplemented!(),
+        SudoMsg::VerifyUpgradeAndUpdateState(_) => unimplemented!(),
+        SudoMsg::MigrateClientStore(_) => unimplemented!(),
+    };
+
+    Ok(Response::default().set_data(result))
+}
+
+pub fn verify_membership(
+    deps: Deps<EthereumCustomQuery>,
+    verify_membership_msg: VerifyMembershipMsg,
+) -> Result<Binary, ContractError> {
+    let eth_client_state = get_eth_client_state(deps);
+    let eth_consensus_state = get_eth_consensus_state(deps, &verify_membership_msg.height);
+
+    ethereum_light_client::membership::verify_membership(
+        eth_consensus_state,
+        eth_client_state,
+        verify_membership_msg.proof.into(),
+        verify_membership_msg
+            .merkle_path
+            .key_path
+            .into_iter()
+            .map(Into::into)
+            .collect(),
+        Some(verify_membership_msg.value.into()),
+    )
+    .map_err(ContractError::VerifyMembershipFailed)?;
+
+    Ok(to_json_binary(&Ok::<(), ()>(()))?)
+}
+
+pub fn verify_non_membership(
+    deps: Deps<EthereumCustomQuery>,
+    verify_non_membership_msg: VerifyNonMembershipMsg,
+) -> Result<Binary, ContractError> {
+    let eth_client_state = get_eth_client_state(deps);
+    let eth_consensus_state = get_eth_consensus_state(deps, &verify_non_membership_msg.height);
+
+    ethereum_light_client::membership::verify_membership(
+        eth_consensus_state,
+        eth_client_state,
+        verify_non_membership_msg.proof.into(),
+        verify_non_membership_msg
+            .merkle_path
+            .key_path
+            .into_iter()
+            .map(Into::into)
+            .collect(),
+        None,
+    )
+    .map_err(ContractError::VerifyNonMembershipFailed)?;
+
+    Ok(to_json_binary(&Ok::<(), ()>(()))?)
+}
+
+pub fn update_state() -> Result<Binary, ContractError> {
+    Ok(to_json_binary(&UpdateStateResult { heights: vec![] })?)
+}
```

Comment:
These should be moved under a submodule for better readability:
```rs
mod sudo {
  // ...
}
```
and then
```rs
pub fn sudo(
    deps: DepsMut<EthereumCustomQuery>,
    _env: Env,
    msg: SudoMsg,
) -> Result<Response, ContractError> {
    let result = match msg {
        SudoMsg::VerifyMembership(verify_membership_msg) => {
            sudo::verify_membership(deps.as_ref(), verify_membership_msg)?
        }
        SudoMsg::VerifyNonMembership(verify_non_membership_msg) => {
            sudo::verify_non_membership(deps.as_ref(), verify_non_membership_msg)?
        }
// ...
```

Reply:
Agreed. Will do that in #143